woocommerce/woocommerce-ios
# Order List → Support Future Orders

Refs #956. This adds support for Future Orders (p99K0U-1X8-p2). 

> “Future orders” are orders that have a date_created set for sometime in the future. 

Previously, we were always [excluding orders that were created a day after today](https://github.com/woocommerce/woocommerce-ios/blob/9f9801005d484ce25b6646449789037afb2ec53e/WooCommerce/Classes/ViewRelated/Orders/OrdersViewModel.swift#L44-L47). Now, the Orders tabs work like this:

Processing | All Orders
--------|-------
![image](https://user-images.githubusercontent.com/198826/79592368-21943700-8097-11ea-8988-26983963d7c9.png)        |       ![image](https://user-images.githubusercontent.com/198826/79592267-fc9fc400-8096-11ea-9ce0-f77ed003c9a1.png)
Includes all orders regardless of when they were created | Only includes orders up to **midnight of today**
Groups orders from tomorrow up to the end of time in "Upcoming" section | 

All other filtered lists like Orders → Search → Completed will show future orders. It is only the All Orders list that does not show them. 

## How It Works

### Configuration 

Since the order list classes are configurable, I added an `includeFutureOrders` property in `OrdersListViewModel`: 

https://github.com/woocommerce/woocommerce-ios/blob/ccb9f3cdb9c2d1754fd6909afb52dd27cdf7ad0a/WooCommerce/Classes/ViewRelated/Orders/OrdersViewModel.swift#L31-L37

This property is now used in the `ResultsController` query. It is `true` by default but it is set to `false` in the All Orders tab:

https://github.com/woocommerce/woocommerce-ios/blob/ccb9f3cdb9c2d1754fd6909afb52dd27cdf7ad0a/WooCommerce/Classes/ViewRelated/Orders/OrdersMasterViewController.swift#L93-L96

### Section Grouping

We use the [`normalizedAgeAsString`](https://github.com/woocommerce/woocommerce-ios/blob/ccb9f3cdb9c2d1754fd6909afb52dd27cdf7ad0a/WooCommerce/Classes/ViewRelated/Orders/OrdersViewModel.swift#L44-L47) to group the sections. It uses the `Age` enum underneath so in order to add the "Upcoming" grouping, I had to add a `.upcoming` option:

https://github.com/woocommerce/woocommerce-ios/blob/ccb9f3cdb9c2d1754fd6909afb52dd27cdf7ad0a/WooCommerce/Classes/Model/Age.swift#L12

The calculation for the grouping is in [`Age.from()`](https://github.com/woocommerce/woocommerce-ios/blob/ccb9f3cdb9c2d1754fd6909afb52dd27cdf7ad0a/WooCommerce/Classes/Model/Age.swift#L31-L33). I didn't see any existing unit tests for it so I added some in e300802e58714b50e6d8432f0bee4ca5fba837c5 before I added the `.upcoming` changes to make sure I didn't break anything. 

### REST API Arguments

Unfortunately, the way the `SyncCoordinator` works is that the `ResultsController` fetching predicates **and** the API fetching predicates have to **exactly match** in order for the infinite paging mechanism to work. I'm not quite sure if this is a good constraint since it was so easy to break. 

Anyway, I added a parameter in the `OrdersRemote.loadAllOrders` in order to make the paging work again:

https://github.com/woocommerce/woocommerce-ios/blob/ccb9f3cdb9c2d1754fd6909afb52dd27cdf7ad0a/Networking/Networking/Remote/OrdersRemote.swift#L14-L22

The value is ultimately passed from the `OrdersViewModel`:

https://github.com/woocommerce/woocommerce-ios/blob/ccb9f3cdb9c2d1754fd6909afb52dd27cdf7ad0a/WooCommerce/Classes/ViewRelated/Orders/OrdersViewModel.swift#L164

## Date.tomorrow() Bug Fix

I found out that the `Date.tomorrow()` method we used to filter for `dateCreated` didn't work as expected because it still returns the **time**. The value may become tomorrow + sometime in the afternoon. This results in us still showing future orders in All Orders. I fixed this by creating a new `Date.nextMidnight()` method in 9deddf09c2492c237c36a944114eeb8b808d2f91. 

We're now using this method in the `ResultsController` fetching and in the REST API fetching. 

## Missing Unit Tests

I added quite a few unit tests in this. I think they're even bigger than the actual changes. 🤪 I didn't add unit tests for the `OrderStore` and `OrdersRemote` changes because this PR was close to 500 already. I'll add those in another PR. 

## Testing

Please review p99K0U-1X8-p2 for the spec of Future Orders to make sure I didn't miss anything. 

1. In wp-admin, create orders with creation date set in the future. Please try to create orders with different statuses like Processing, Completed, etc. 
2. Navigate to Orders → Processing. 
3. Pull to refresh. 
3. Confirm that there is an Upcoming section with the processing orders that you created. 
3. Confirm that paging works. 
3. Navigate to Orders → All Orders. 
4. Confirm that there is no Upcoming section and that there are no shown future orders. 
3. Confirm that paging works. 
4. Navigate to Orders → Search → [any filter]. 
5. Confirm that there is an Upcoming section with the future orders that you created. 

## Reviewing 

Only 1 reviewer is needed but anyone can review. 

## Submitter Checklist
